server/src/main/java/io/crate/execution/dml/upsert/FromRawInsertSource.java

```diff
@@ -21,27 +21,148 @@
+        try {
+            var parser = JsonXContent.JSON_XCONTENT.createParser(
+                NamedXContentRegistry.EMPTY,
+                DeprecationHandler.THROW_UNSUPPORTED_OPERATION,
+                new BytesArray(src).array());
+            parsed = parser.map();
+            if (parser.getTokenLocation().columnNumber < src.length()) {
+                throw new IllegalArgumentException("failed to parse `" + src.substring(parser.getTokenLocation().columnNumber) + "`");
+            }
+        } catch (IOException e) {
+            if (e instanceof JsonParseException) {
+                throw new IllegalArgumentException("failed to parse");
+            }
+        }
+
+        Map<String, Object> validatedSource = validateRawSource(parsed);
+        mixinDefaults(validatedSource,
+                      // evaluateDefaultValues is invoked per row for expressions like now(), and gen_random_text_uuid().
+                      // Other deterministic expressions like abs(-10) should have been evaluated once and for all by the analyzer.
+                      evaluateDefaultValues(table.defaultExpressionColumns(), txnCtx, nodeCtx));
+
+        GeneratedColumns<Map<String, Object>> generatedColumns = new GeneratedColumns<>(
+            inputFactory,
+            txnCtx,
+            FromSourceRefResolver.WITHOUT_PARTITIONED_BY_REFS,
+            refLookUpCache.values(),
+            table.generatedColumns());
+
+        generatedColumns.setNextRow(validatedSource);
+        generatedColumns.validateValues(validatedSource);
+
+        for (var entry : generatedColumns.generatedToInject()) {
+            var reference = entry.getKey();
+            var value = entry.getValue().value();
+            var valueForInsert = reference
+                .valueType()
+                .valueForInsert(value);
+            var column = reference.column();
+            Maps.mergeInto(validatedSource, column.name(), column.path(), valueForInsert);
+        }
+
+        checks.validate(validatedSource);
+
+        return validatedSource;
+    }
+
+    private Map<String, Object> validateRawSource(Map<String, Object> rawSource) {
+        Map<String, Object> casted = new HashMap<>(rawSource.size());
+        for (var entry : rawSource.entrySet()) {
+            String srcCol = entry.getKey();
+            Object data = entry.getValue();
+            Reference ref = refLookUpCache.get(srcCol);
+            if (ref == null) {
+                try {
+                    ref = table.resolveColumn(srcCol, true, txnCtx.sessionSettings().errorOnUnknownObjectKey());
+                } catch (ColumnUnknownException e) {
+                    // This handles unknown columns of strict tables and tables only.
+                    // Unknown subcolumns of strict objects will be created here and the validation happens at DocumentMapper
+                    throw new StrictDynamicMappingException(table.ident().toString(), srcCol);
```

Comment:
Not sure I understand this comment. If the `column_policy` is strict, implicit dynamic column creation should be prohibited. 
Shouldn't we let the original `ColumnUnknownException` bubble up?